### DISCUSSION TITLE
Make PoCL self-contained (by relying less on external binaries)

### DISCUSSION THREAD

**Discussion by maleadt** [pocl/pocl#1741] [Ideas]
I'm using PoCL as a shared library in an environment where I don't control anything on the system _but_ a user-writable folder where I put downloaded libraries and binaries. The intended workflow is:

```julia-repl
# user downloads, installs and opens Julia

julia> using Pkg
julia> Pkg.add(["OpenCL", "pocl])
       # this downloads a precompiled OpenCL ICD loader as well as PoCL and
       # all of its dependencies somewhere in `~/.julia`

julia> using OpenCL, pocl
       # here, we actually dlopen all these libraries. pocl is registered with
       # the OpenCL ICD by populating `OCL_ICD_FILENAMES` in the environment.

julia> # the user can now use PoCL through OpenCL.jl
```

After some weeks of hacking this works relatively OK. The biggest hurdle was that PoCL is not actually a fully self-contained shared library in that it executes many binaries during kernel compilation. This has several drawbacks:

- binaries do not automatically find library dependencies in `~/.julia`, so we have to use wrapper scripts that populate `LD_LIBRARY_PATH`/`DYLD_LIBRARY_PATH`/`PATH` (this isn't a problem with libraries as we eagerly `dlopen` dependencies before loading a library, making sure the dynamic loader can find them)
- tools like `clang` are themselves not self-contained, e.g., looking up system libraries like `libgcc_s.so` or `crt0.so` which we now have to copy along and force Clang to use by passing `-L`, or executing the system linker for which we redistribute `lld` now and use through `clang -fuse-ld=lld --ld-path=...`.

These hacks were made possible env vars introduced in https://github.com/pocl/pocl/pull/1543 and https://github.com/pocl/pocl/pull/1625.

---

Meanwhile, vendor OpenCL implementations (or similar libraries) are typically very self-contained, compiling and executing kernel code without much more. This is true for NVIDIA's `libnvvm` (consuming CUDA C code), NVIDIA's `libnvptxcompiler` (consuming PTX assembly; even distributed as a simple static library), Intel's `libocloc` (consuming both OpenCL C and SPIR-V), etc.

I think it would be really valuable if PoCL behaves more like this as well.
- `llvm-spirv` could be replaced by its library, providing `readSpirv` to translate to LLVM IR
- `clang` is already mostly used through `libclang`, but PoCL is still invoking the main driver. Maybe it's better to take more control of the compilation a la https://blog.memzero.de/libclang-c-to-llvm-ir/ to avoid pulling in system stuff
- to avoid having to ever link using `clang` and the system linker, PoCL could maybe save just the executable code and load it through LLVM's JIT capabilities

I realize these are big ticket items, but nonetheless I think it would be very valuable if it were possible to distribute `libpocl.so` without anything more, and have it behave correctly regardless of the host system.

x-ref https://github.com/JuliaGPU/OpenCL.jl/issues/259

**Comment by franz** [comment]
Hello,

Just one thing immediately coming to mind:

* llvm-spirv could be replaced by its library, providing readSpirv to translate to LLVM IR

... this is at least partially implemented. PoCL is able to use `libLLVMSPIRV` library instead, but IIRC not all code paths have been converted to be able to use both the binary and library. This might be relatively easy to finish; the problem is that `readSpirv` and `writeSpirv` are not enough. We need the something to extract & change Spec Constants in the SPIR-V (for `clSetProgramSpecializationConstant`). I have not checked if  `libLLVMSPIRV` provides anything useful for that case.

**Comment by pjaaskel** [comment]
I very much agree. We have come a long way since the beginning of PoCL 14 years or so ago when we called also "opt" and llc via shell scripts. Now at least those phases are invoked via library APIs (with the drawback of the unstable LLVM C++ API). Unfortunately due to small corner use cases such as @franz pointed out it might not be straightforward to go all the way, but I'd be very happy to see it happen eventually!

**Comment by maleadt** [comment]
> We need the something to extract & change Spec Constants in the SPIR-V (for `clSetProgramSpecializationConstant`). I have not checked if `libLLVMSPIRV` provides anything useful for that case.

`readSpirv` takes an options parameter which supports setting specialization constants: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/50af2c70832745be1c0ce1562c4038162c319e5c/include/LLVMSPIRVOpts.h#L175-L177
